proof/crefine/lib/Boolean_C.thy

```diff
@@ -0,0 +1,140 @@
+(*
+ * Copyright 2023, Proofcraft Pty Ltd
+ * Copyright 2014, General Dynamics C4 Systems
+ *
+ * SPDX-License-Identifier: GPL-2.0-only
+ *)
+
+(* Definitions and rules for C boolean constants *)
+
+theory Boolean_C
+imports
+  "CSpec.KernelInc_C"
+begin
+
+lemma true_and_1[simp]:
+  "true && 1 = true"
+  "signed true && 1 = signed true"
+  "unsigned true && 1 = unsigned true"
+  by (simp add: true_def)+
+
+lemma to_bool_true[simp]:
+  "to_bool true"
+  "to_bool (signed true)"
+  "to_bool (unsigned true)"
+  by (simp add: to_bool_def true_def)+
+
+lemma true_equals_simps[simp]:
+  "true = 1 \<longleftrightarrow> True"
+  "signed true = 1 \<longleftrightarrow> True"
+  "unsigned true = 1 \<longleftrightarrow> True"
+  "true = 0 \<longleftrightarrow> False"
+  "signed true = 0 \<longleftrightarrow> False"
+  "unsigned true = 0 \<longleftrightarrow> False"
+  by (simp add: true_def)+
+
+lemma true_eq_from_bool[simp]:
+  "(true = from_bool P) = P"
+  "(signed true = from_bool P) = P"
+  "(unsigned true = from_bool P) = P"
+  by (simp add: from_bool_def split: bool.splits)+
+
+lemma false_and_1[simp]:
+  "false && 1 = false"
+  "signed false && 1 = signed false"
+  "unsigned false && 1 = unsigned false"
+  by (simp add: false_def)+
+
+lemma to_bool_false[simp]:
+  "\<not> to_bool false"
+  "\<not> to_bool (signed false)"
+  "\<not> to_bool (unsigned false)"
+  by (simp add: to_bool_def false_def)+
+
+lemma false_equals_simps[simp]:
+  "false = 0 \<longleftrightarrow> True"
+  "signed false = 0 \<longleftrightarrow> True"
+  "unsigned false = 0 \<longleftrightarrow> True"
+  "false = 1 \<longleftrightarrow> False"
+  "signed false = 1 \<longleftrightarrow> False"
+  "unsigned false = 1 \<longleftrightarrow> False"
+  by (simp add: false_def)+
+
+lemma false_eq_from_bool[simp]:
+  "(false = from_bool P) = (\<not> P)"
+  "(signed false = from_bool P) = (\<not> P)"
+  "(unsigned false = from_bool P) = (\<not> P)"
+  by (simp add: from_bool_def split: bool.splits)+
+
+lemma from_bool_vals[simp]:
+  "from_bool True = signed true"
+  "from_bool False = signed false"
+  by (simp add: from_bool_def)+
+
+lemma true_neq_false[simp]:
+  "true \<noteq> false"
+  "signed true \<noteq> signed false"
+  "unsigned true \<noteq> unsigned false"
+  by (simp add: true_def false_def)+
+
+(* this is slightly weird, but the bitfield generator
+   masks everything with the expected bit length.
+   So we do that here too. *)
```

Comment:
was this the original comment?

Reply:
Yeah, just copied across from `Wellformed_C`

Reply:
Hmm, could we tweak it to either be more content-oriented rather than feelings-of-writer-oriented and/or fitting on fewer lines?